### PR TITLE
let :Launch/Open use autoload functions

### PR DESCRIPTION
**Pull request by Konfekt**
addresses https://github.com/vim/vim/issues/15959

### CODE REVIEW

File: runtime/autoload/netrw.vim [hunk 1]
```diff
@@ -5294,6 +5294,104 @@ fun! s:NetrwBrowseUpDir(islocal)
...
+    else
+      return &srr .. (has("win32") ? "nul" : "/dev/null")
+    endif
+  endif
+  return ''
+endfunc
+
+fun! netrw#Launch(args)
+   let args = trim(a:args)
+   if has('unix')
```
Comment (Konfekt) [inline runtime/autoload/netrw.vim]: It's unnecessary to check for the OS on every invocation, but too much of a hassle to tear the checks apart out of the function

Comment (zzzyxwvut) [reply runtime/autoload/netrw.vim]: Platform tests can be pushed all the way up out of function  
bodies and a separate `netrw#Launch` implementation written  
for each supported case.

```diff
diff --git a/runtime/autoload/netrw.vim b/runtime/autoload/netrw.vim
index 50650ae58..ecdab8bb3 100644
--- a/runtime/autoload/netrw.vim
+++ b/runtime/autoload/netrw.vim
@@ -5307,38 +5307,52 @@ func s:redir()
   return ''
 endfunc
 
-fun! netrw#Launch(args)
-   let args = trim(a:args)
-   if has('unix')
-      if has('win32unix')
-         " If cygstart provided, then assume Cygwin and use `cygstart --hide`; see man cygstart.
-         if executable('cygstart')
-            exe 'silent ! cygstart --hide' args s:redir() | redraw!
-         elseif !empty($MSYSTEM) && executable('start')
-            " MSYS2/Git Bash comes by default without cygstart; see
-            " https://www.msys2.org/wiki/How-does-MSYS2-differ-from-Cygwin
-            " Instead it provides /usr/bin/start script running `cmd.exe //c start`
-            " Adding "" //b` sets void title, hides cmd window and blocks path conversion
-            " of /b to \b\ " by MSYS2; see https://www.msys2.org/docs/filesystem-paths/
-            exe 'silent !start "" //b' args  s:redir() | redraw!
-         else
-            " imitate /usr/bin/start script for other environments and hope for the best
-            exe 'silent !cmd //c start "" //b' args  s:redir() | redraw!
-         endif
-      elseif exists('$WSL_DISTRO_NAME') " use cmd.exe to start GUI apps in WSL
-         exe 'silent !'..
-                  \ ((args =~? '\v<\f+\.(exe|com|bat|cmd)>') ?
-                  \ 'cmd.exe /c start "" /b ' .. args :
-                  \ 'nohup ' .. args .. ' ' .. s:redir() .. '&')
-                  \ | redraw!
-      else
-         exe ':silent ! nohup' args s:redir() '&' | redraw!
-      endif
-   elseif has('win32')
-      exe 'silent !'.. (&shell =~? '\<cmd\.exe\>' ? '' : 'cmd.exe /c')
-               \ 'start /b ' args s:redir() | redraw!
-   endif
-endfun
+if has('unix')
+ if has('win32unix')
+  " If cygstart provided, then assume Cygwin and use `cygstart --hide`; see man cygstart.
+  if executable('cygstart')
+   fun! netrw#Launch(args)
+     exe 'silent ! cygstart --hide' trim(a:args) s:redir() | redraw!
+   endfun
+  elseif !empty($MSYSTEM) && executable('start')
+   " MSYS2/Git Bash comes by default without cygstart; see
+   " https://www.msys2.org/wiki/How-does-MSYS2-differ-from-Cygwin
+   " Instead it provides /usr/bin/start script running `cmd.exe //c start`
+   " Adding "" //b` sets void title, hides cmd window and blocks path conversion
+   " of /b to \b\ " by MSYS2; see https://www.msys2.org/docs/filesystem-paths/
+   fun! netrw#Launch(args)
+     exe 'silent !start "" //b' trim(a:args) s:redir() | redraw!
+   endfun
+  else
+   " imitate /usr/bin/start script for other environments and hope for the best
+   fun! netrw#Launch(args)
+     exe 'silent !cmd //c start "" //b' trim(a:args) s:redir() | redraw!
+   endfun
+  endif
+ elseif exists('$WSL_DISTRO_NAME') " use cmd.exe to start GUI apps in WSL
+  fun! netrw#Launch(args)
+    let args = trim(a:args)
+    exe 'silent !' ..
+      \ ((args =~? '\v<\f+\.(exe|com|bat|cmd)>') ?
+      \ 'cmd.exe /c start "" /b ' .. args :
+      \ 'nohup ' .. args .. ' ' .. s:redir() .. '&')
+      \ | redraw!
+  endfun
+ else
+  fun! netrw#Launch(args)
+    exe ':silent ! nohup' trim(a:args) s:redir() '&' | redraw!
+  endfun
+ endif
+elseif has('win32')
+ fun! netrw#Launch(args)
+   exe 'silent !' .. (&shell =~? '\<cmd\.exe\>' ? '' : 'cmd.exe /c')
+     \ 'start /b ' trim(a:args) s:redir() | redraw!
+ endfun
+else
+ fun! netrw#Launch(dummy)
+   echom 'No common launcher found'
+ endfun
+endif
 
 if exists('g:netrw_browsex_viewer') && executable(g:netrw_browsex_viewer)
    " extract any viewing options.  Assumes that they're set apart by spaces.
```

(Unfortunately, a mix of indentation styles has accreted over  
the years.  At one time (`v7.0131`), the style used to be `sw=1`  
and an extra leading space character for function bodies and  
line continuation, and using no tabs for indentation.)

Comment (Konfekt) [reply runtime/autoload/netrw.vim]: > Platform tests can be pushed all the way up out of function
bodies and a separate `netrw#Launch()` implementation written
for each supported case.

Oh, thanks for the patch!

> Unfortunately, a mix of indentation styles has accreted over
the years. At one time (v7.0131), the style used to be sw=1
and an extra leading space character for function bodies and
line continuation, and using no tabs for indentation.)

It's a mess. The file should also be split

**Comment by lifepillar** [comment]
I've tested this and I confirm that it fixes https://github.com/vim/vim/issues/15959. Great job, thanks!

**Comment by Konfekt** [comment]
@lifepillar : could you test again after the clean up?

**Comment by Konfekt** [comment]
How about adding a short form `:O` (similar to `:E`) ?

**Comment by chrisbra** [comment]
> How about adding a short form :O (similar to :E) ?

Let's not waste such things for users. It's easy enough to add custom short commands for users, but if we now use `:O` as a short form, this means users cannot re-use `:O` themselves for something different. `:Open` is fine and clear from the name what it does, `:O` not so much.

**Comment by chrisbra** [comment]
@Konfekt Sorry, for not crediting properly, I forgot to use `git commit --author` this time.

**Comment by Konfekt** [comment]
No problem at all. 

I also agree with your take on `:O` as I despise these short names popular among recent command-line utilities polluting these few-letter namespace of user's aliases. I suppose then that `:E` counts for damage done. 

In any event, here are some [usage examples](https://old.reddit.com/r/vim/comments/1ggahwr/open_path_launch_app/) for these newly included commands, in case it helps to have some in the docs

**Comment by your-diary** [comment]
Is it possible to NOT define `Open` command introduced in this PR?
(Some configurable option?)

This PR breaks my existing `~/.vimrc`, in which my own `Open` command is defined.

```
$ vi
Error detected while processing /opt/homebrew/Cellar/vim/9.1.0850/share/vim/vim91/plugin/netrwPlugin.vim:
line   39:
E174: Command already exists: add ! to replace it: Open    call netrw#Open(trim(<q-args>))
```

**Comment by Konfekt** [comment]
Yes, I also defined my proper `:Open` command previously. By `:help rtp`, your commands load first, so you'd have to check for something like the version number by

```vim
if v:version > 901 || (v:version == 901 && has('patch-9.1.0822')) && get(g:, 'loaded_netrwPlugin', 0)
  command -complete=file     -nargs=1   Open    call MyOpen(<q-args>)
endif
```

**Comment by your-diary** [comment]
@Konfekt 
Doesn't your code mean to define your own version of `Open` only for older versions of Vim?

My question is: how to prevent `netrw` from defining `Open`? I do want to use my own `Open`, which I've used for years.

**Comment by Konfekt** [comment]
Yes, this was meant. If you want to override it by your own command, use `command!` with a bang in a folder read after `$VIMRUNTIME`, say `~/.vim/after/plugin` if you're on Linux or `%USERPROFILE%/vimfiles/after/plugin` on Microsoft Windows

**Pull request closed without merging**